File: translated/f-to-c/DRB060-matrixmultiply-orig-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Classic i-k-j matrix multiplication. No data race pairs.

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main()
{
    int N, M, K, len, i, j, l;
    float** a;
    float** b;
    float** c;

    len = 100;
    N = len;
    M = len;
    K = len;

    // Allocate 2D arrays
    a = (float**)malloc(N * sizeof(float*));
    b = (float**)malloc(M * sizeof(float*));
    c = (float**)malloc(K * sizeof(float*));
    for (i = 0; i < N; i++) {
        a[i] = (float*)malloc(M * sizeof(float));
    }
    for (i = 0; i < M; i++) {
        b[i] = (float*)malloc(K * sizeof(float));
    }
    for (i = 0; i < K; i++) {
        c[i] = (float*)malloc(N * sizeof(float));
    }

    #pragma omp parallel for private(j, l)
    for (i = 0; i < N; i++) {
        for (l = 0; l < K; l++) {
            for (j = 0; j < M; j++) {
                c[i][j] = c[i][j] + a[i][l] * b[l][j];
            }
        }
    }

    // Free 2D arrays
    for (i = 0; i < N; i++) {
        free(a[i]);
    }
    for (i = 0; i < M; i++) {
        free(b[i]);
    }
    for (i = 0; i < K; i++) {
        free(c[i]);
    }
    free(a);
    free(b);
    free(c);

    return 0;
} 

File: translated/f-to-go/DRB060-matrixmultiply-orig-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Classic i-k-j matrix multiplication. No data race pairs.

package main

import (
	"runtime"
	"sync"
)

func main() {
	var N, M, K, length, i int
	var a, b, c [][]float32

	length = 100
	N = length
	M = length
	K = length

	// Allocate matrices
	a = make([][]float32, N)
	b = make([][]float32, M)
	c = make([][]float32, K)
	for i = 0; i < N; i++ {
		a[i] = make([]float32, M)
	}
	for i = 0; i < M; i++ {
		b[i] = make([]float32, K)
	}
	for i = 0; i < K; i++ {
		c[i] = make([]float32, N)
	}

	//$omp parallel do private(j, l)
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := N / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= N; start += chunkSize {
		end := start + chunkSize - 1
		if end > N {
			end = N
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				for l := 1; l <= K; l++ { // j and l are private to each goroutine
					for j := 1; j <= M; j++ {
						c[i-1][j-1] = c[i-1][j-1] + a[i-1][l-1]*b[l-1][j-1] // No race - proper parallelization
					}
				}
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end parallel do

	// deallocate(a,b,c)
}
